llvm/include/llvm/CodeGen/TargetLowering.h

```diff
@@ -4339,6 +4339,21 @@ class TargetLowering : public TargetLoweringBase {
     return isTypeLegal(VT);
   }
 
+  /// Same as isTypeDesirableForOp(unsigned Opc, EVT VT), but also check if
+  /// the target is 'desirable' to truncate or extend OldVT to NewVT only using
+  /// the given node type, without the need of explicit trunc or ext. e.g. On
+  /// RISC-V Vector extension, vnsrl.wi can directly convert <n x i32> to <n x
+  /// i16> when shifting, with no extra trunc operations needed.
+  virtual bool isTypeDesirableForOpWithCast(unsigned Opc, EVT NewVT,
+                                            EVT OldVT) const {
```

Comment:
Did you investigate whether we could just update the existing isTypeDesirableForOp? The x86 changes would be pretty easy and I suspect the AMDGPU changes would be as well

Reply:
There are tons of usage of isTypeDesirableForOp, we need to make change everywhere if the parameter changes. Besides, as far as I know, it's rare for instructions to truncate operand while doing other jobs. Updating the existing one to NewVT + OldVT may be unnecessary for most of the cases.

Reply:
Nearly all of those isTypeDesirableForOp calls are testing whether we should promote/truncate an existing op with the new type - including the old type as well is likely to be useful.

Reply:
I looked through the code and realized isTypeDesirableForOp is not suitable for the issue. isTruncateFree(SDValue, EVT) seems better here.